File: mobile/lib/features/architect/domains/use_cases/team/leave.dart
import 'package:dartz/dartz.dart';
import 'package:equatable/equatable.dart';

import '../../../../../core/errors/failure.dart';
import '../../../../../core/use_cases/usecase.dart';
import '../../entities/team.dart';
import '../../repositories/team.dart';

class TeamLeave implements UseCase<Team, Params> {
  final TeamRepository repository;

  TeamLeave(this.repository);

  @override
  Future<Either<Failure, Team>> call(Params params) async {
    return await repository.leave(params.teamId, params.userId);
  }
}

class Params extends Equatable {
  final String teamId;
  final String userId;

  const Params({
    required this.teamId,
    required this.userId
  });

  @override
  List<Object?> get props => [teamId, userId];
}

File: backend/api/app/domain/use_cases/team/leave.py
from core.use_cases.use_case import UseCase
from app.domain.repositories.team import BaseRepository
from core.common.equatable import Equatable
from core.common.either import Either
from core.errors.failure import Failure
from app.domain.entities.team import TeamEntity

class Params(Equatable):
    def __init__(self, team_id: str, user_id: str) -> None:
        self.team_id = team_id
        self.user_id = user_id

class LeaveTeam(UseCase[TeamEntity]):
    def __init__(self, repository: BaseRepository):
        self.repository = repository
    
    async def __call__(self, params: Params) -> Either[Failure, TeamEntity]:
        return await self.repository.leave_team(params.team_id, params.user_id)

